command/providers.go

```diff
@@ -94,46 +94,60 @@ func (c *ProvidersCommand) Run(args []string) int {
 		return 1
 	}
 
-	reqs, reqDiags := config.ProviderRequirements()
-	if reqDiags.HasErrors() {
-		c.showDiagnostics(configDiags)
+	reqs, reqDiags := config.ProviderRequirementsByModule()
+	diags = diags.Append(reqDiags)
+	if diags.HasErrors() {
+		c.showDiagnostics(diags)
 		return 1
 	}
 
 	state := s.State()
+	var stateReqs getproviders.Requirements
 	if state != nil {
-		stateReqs := state.ProviderRequirements()
-		reqs = reqs.Merge(stateReqs)
+		stateReqs = state.ProviderRequirements()
 	}
 
 	printRoot := treeprint.New()
-	providersCommandPopulateTreeNode(printRoot, reqs)
+	c.populateTreeNode(printRoot, reqs)
 
+	c.Ui.Output("\nProviders required by configuration:")
 	c.Ui.Output(printRoot.String())
 
+	if len(stateReqs) > 0 {
+		c.Ui.Output("Providers required by state:\n")
+		for fqn := range stateReqs {
+			c.Ui.Output(fmt.Sprintf("    provider[%s]\n", fqn.String()))
+		}
+	}
+
 	c.showDiagnostics(diags)
 	if diags.HasErrors() {
 		return 1
 	}
 	return 0
 }
 
-func providersCommandPopulateTreeNode(node treeprint.Tree, deps getproviders.Requirements) {
-	for fqn, dep := range deps {
+func (c *ProvidersCommand) populateTreeNode(tree treeprint.Tree, node *configs.ModuleRequirements) {
```

Comment:
Was there a specific reason or benefit to make this a method on the `ProvidersCommand`? I'm not opposed per se, but I don't see it using anything from the `ProvidersCommand` to explain the change - without this other commands could call this function.

Reply:
Ah, no good reason. I'm happy to change it back if you prefer!

In prototyping this feature I added another function to this file, and since it was also very providers-command specific I named it `providersCommandWhatever` as well. Then that made me realize that if we want to signal that the purpose of the function is specific to this command, making it a method is probably clearer than a long name prefix, so I did that for both. Then I removed that function and forgot about this one.

I think we have both patterns elsewhere in the command package: methods which don't reference the receiver, and package-scoped functions which are really only related to one struct. I slightly prefer the former but only at a 2/5 level.